src/Generator/Generators/CSharp/CSharpTextTemplate.cs

```diff
@@ -1273,10 +1273,51 @@ public void GenerateVTable(Class @class)
 
             GenerateVTableClassSetup(@class, wrappedEntries);
 
+            GenerateVTableCleanup(@class, wrappedEntries);
+
             WriteLine("#endregion");
             PopBlock(NewLineKind.BeforeNextBlock);
         }
 
+        private void GenerateVTableCleanup(Class @class, IList<VTableComponent> wrappedEntries)
+        {
+            WriteLine("private static readonly Destructor Finalise = new Destructor();");
+            WriteLine("private sealed class Destructor");
+            WriteStartBraceIndent();
+            WriteLine("~Destructor()");
+            WriteStartBraceIndent();
+            WriteLine("if (__ManagedVTables != null)");
+            WriteStartBraceIndent();
+            DeAllocateVTables(@class, wrappedEntries, false);
+            WriteCloseBraceIndent();
+            WriteCloseBraceIndent();
+            WriteCloseBraceIndent();
+        }
+
+        private void DeAllocateVTables(Class @class, IList<VTableComponent> wrappedEntries,
+            bool destructorOnly)
+        {
+            if (Options.IsMicrosoftAbi)
+                DeAllocateVTablesMS(@class, wrappedEntries, destructorOnly);
+            else
+                DeAllocateVTablesItanium(@class, wrappedEntries, destructorOnly);
+        }
+
+        private void DeAllocateVTablesMS(Class @class, IList<VTableComponent> wrappedEntries,
+            bool destructorOnly)
+        {
+            var managedVTables = destructorOnly ? "__ManagedVTablesDtorOnly" : "__ManagedVTables";
```

Comment:
The logic here is not the same as the patching logic. In here both managed tables might be full so you need to check each one and if not null, destroy it - not one or the other as it is at the moment.